core/src/main/java/org/apache/iceberg/util/EnvironmentUtil.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.iceberg.util;
+
+import java.util.Collections;
+import java.util.Map;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+
+public class EnvironmentUtil {
+  private EnvironmentUtil() {
+  }
+
+  private static final String ENVIRONMENT_VARIABLE_PREFIX = "env:";
+
+  public static Map<String, String> resolveAll(Map<String, String> properties) {
+    return resolveAll(System.getenv(), properties);
+  }
+
+  @VisibleForTesting
+  static Map<String, String> resolveAll(Map<String, String> env, Map<String, String> properties) {
+    Map<String, String> result = Maps.newHashMap();
+    properties.forEach((name, value) -> {
+      if (value != null && value.startsWith(ENVIRONMENT_VARIABLE_PREFIX)) {
+        result.put(name, env.get(value.substring(ENVIRONMENT_VARIABLE_PREFIX.length())));
+      } else {
+        result.put(name, value);
+      }
+    });
+
+    return Collections.unmodifiableMap(result);
```

Comment:
I don't know that we modify the property map anywhere in the RESTCatalog, but this wasn't a requirement perviously.

Reply:
We currently build the property maps as immutable maps in several places. And that brings up an issue where if this produces a null key, those areas will fail. Instead I'm going to change this so that missing environment variables result in not adding anything to the map, and will switch to an `ImmutableMap` builder.